cluster-autoscaler/cloudprovider/openshiftmachineapi/machineapi_nodegroup.go

```diff
@@ -111,17 +126,21 @@ func (ng *nodegroup) DeleteNodes(nodes []*apiv1.Node) error {
 		}
 
 		machine.Annotations[machineDeleteAnnotationKey] = time.Now().String()
-
 		if _, err := ng.machineapiClient.Machines(machine.Namespace).Update(machine); err != nil {
-			return fmt.Errorf("failed to update machine %s/%s: %v", machine.Namespace, machine.Name, err)
+			return err
+		}
```

Comment:
should we wait here till machine deletion? And handle machine deletion failure scenario here before updating scalableResource? or atleast ensure here that machine's deletiontimestamp has gone non-nil?

Reply:
> Should we wait here till machine deletion?

It needs to be async.

Reply:
> And handle machine deletion failure scenario here before updating scalableResource? or atleast ensure here that machine's deletiontimestamp has gone non-nil?

Are you suggesting something like:
```go
for {
	// get the machine
	// validate machine's timestamp
}
```

the concern I have is how long do we let this run for?

Reply:
yeah, thats what i meant. But now i am thinking it is beyond the scope of the autoscaler. Just annotating successfully should be the max to ensure.
What will be the behavior if scalableResource is updated with lesser replicas after annotating the machine objects and somehow those machine could not get deleted?

Reply:
> What will be the behavior if scalableResource is updated with lesser replicas after annotating the machine objects and somehow those machine could not get deleted?

This is the core of sig-autoscaling's concern; the delete operation is not atomic. In this PR I'm attempting to make it fail-fast trading speed for consistency.